Subscribe support

ikq opened the pull request:
I added Subscriber.js  Notifier.js to JsSIP

I tried write them according  JsSIP style and lint errors.
 Please take a look  and let me know what needs to be fixed.

I already see that *.d.ts files are missing.

The work still in progress: I want add to SUBSCRIBE Contact +sip.instance
as you do in REGISTER

jmillan commented:
Hi @ikq, 

JsSIP code must have the same style in order to make it readable and maintainable. Comments before making a deeper review:

- Debug line as the first method call in al public method implementations.
- New line after return.
- New line before and after call to `super()`.
- First conditional in a block starts in new line and last ends in a new line.
- Conditional blocks always contain curly brackets.
- Comments in the line above when possible.
- Calls to debug have usually a new line above and a new line below.

Please take a certain file (ie: RTCSession.js) as a reference and try to stick to its style.

jmillan commented:
@ikq

Thanks for the cleanup, :+1: We'll review the PR as we can.

jmillan commented:
@ikq,

Cosmetic comment, we use camelCase in API method arguments (not in 100% of the cases, I know, but we will get there). Could you please make such change?

jmillan commented on lib/RequestSender.js:
```diff
@@ -13,7 +13,7 @@ const EventHandlers = {
 module.exports = class RequestSender
 {
-  constructor(ua, request, eventHandlers)
+  constructor(ua, request, eventHandlers, config2=null)
```
I will not check this file right now, because I may refactor this class so it can receive an arbitrary credential.

ikq replied:
if JsSIP will support multiple  (user-name, password, realm) it will be good.
It was workaround.

jmillan commented on lib/UA.js:
```diff
@@ -256,6 +259,39 @@ module.exports = class UA extends EventEmitter
+  /**  
+   * Arbitrary client transaction 
+   */
+  sendRequest(method, target, params, headers, body, handlers, credential)
```
Adding this method here is out of the scope of the PR. New classes should make use of `RequestSender` and `Dialog` internally to send requests.

ikq replied:
I can inline here sendRequest method body.


I don't know how to use Dialog class here (should be checked).
In the Notifier class I  implemented "dialog interface":
dialog.id and dialog.incomingRequest.

jmillan commented on lib/UA.js:
```diff
@@ -256,6 +259,39 @@ module.exports = class UA extends EventEmitter
+  /** 
+   * Create subscriber
+   */
+  subscriber(target, options)
```
Let's use verbs for methods for consistency: `subscribe()`

ikq replied:
OK

jmillan commented on lib/UA.js:
```diff
@@ -256,6 +259,39 @@ module.exports = class UA extends EventEmitter
+  /**
+   * Create notifier
+   */
+  notifier(options)
```
Let's use verbs for methods for consistency: `notify()`

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+    return C;
+  }
+
+  get C()
```
Duplicated method. A static one already defined above.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+  }
+
+  /**
+   * -param {Object} ua reference to JsSIP.UA
```
Let's use JSDoc syntax.

ikq replied:
OK, I'm learning it now...

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+   *   -param {String} allowEvents Allow-Events header value. Optional.
+   *   -param {Object} params Will have priority over ua.configuration. Optional.
+   *      If set please define: to_uri, to_display_name, from_uri, from_display_name
+   *   -param {Array}  headers Optional. Additional SIP headers.
```
Let's rename it to `extraHeaders` .

ikq replied:
OK

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+   *   -param {Array}  headers Optional. Additional SIP headers.
+   *   -param {Object} credential. Will have priority over ua.configuration. Optional.
+   */
+  constructor(ua, target, { eventName, accept, expires, contentType, 
```
As in `Notifier`, let's only put the optional values in the options paramter.

ikq replied:
OK

ikq replied:
OK, the same for Subscriber constructor.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+      throw new TypeError('target is undefined');
+    }
+
+    this._target = target;
```
Move this after the last check.

ikq replied:
OK

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+      throw new TypeError('eventName is undefined');
+    }
+
+    const parsed = Grammar.parse(eventName, 'Event');
```
Move this after the last check.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+      throw new TypeError('eventName - wrong format');
+    }
+    
+    this._event_name = parsed.event;
+    this._event_id = parsed.params && parsed.params.id;
```
Move these after the last check.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,489 @@
+    this._expires = expires;
+    this._allow_events = allowEvents;
+
+    // used to subscribe with body
```
Comments start with Uppercase and end with doc.

ikq replied:
OK. Thanks.

jmillan reviewed (commented):
I've made the first review.

Please go through all the comments and modify as needed. We'll go for sencond review afterwards.

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+    return C;
+  }
+
+  get C()
```
It taken from RTCSession.js  You also use 2 methods.
As I understand:
static method used to internal access
instance method used to access via instance in user code. 
(UA.notifier() expose the class instance)

P.S. in javascript you cannot use static method to access via class instance [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static)

jmillan replied:
It's alright, we should normalize it globally. Not a matter of this PR.

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+   * -param {Object} ua JsSIP UA
+   * -param {Object} options 
+   *   -param {IncomingRequest} subscribe
+   *   -param {String} contentType Content-Type header value
```
Do you mean - NOTIFY Content-Type can be taken from SUBSCRIBE Content-Type or from SUBSCRIBE Accept ?

Hmm...
Accept is optional and contains list of possible NOTIFY content-type RFC 6665 3.1.3.

SUBSCIBE can be without body and Content-Type. 
E.g. in some real server that I use for testing - initial SUBSCRIBE without body.

jmillan replied:
Alright, I thought you where already taking it fromthe subscribe request.

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+   * -param {Object} options 
+   *   -param {IncomingRequest} subscribe
+   *   -param {String} contentType Content-Type header value
+   *   -param {Array}  headers Optional. Additional SIP headers.
```
OK

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+    this._target = subscribe.from.uri.user;
+    subscribe.to_tag = Utils.newTag();
+
+    // NOTIFY request params set according received SUBSCRIBE
```
OK.

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+   * -param {String} reason Optional. To build Subscription-State header
+   * -param {Number} retryAfter Optional. To build Subscription-State header
+   */
+  sendFinalNotify(body = null, reason = null, retryAfter = null) 
```
OK

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+  }
+
+  /**
+   * -param {Object} ua JsSIP UA
```
Question about JSDoc:
I see [https://jsdoc.app/about-getting-started.html](https://jsdoc.app/about-getting-started.html)

/**
 * Represents a book.
 * @constructor
 * @param {string} title - The title of the book.
 * @param {string} author - The author of the book.
 */
function Book(title, author) {
}

In RTCSession you don't describe public methods parameters.
But in UA.js
 /**
   * Make an outgoing call.
   *
   * -param {String} target
   * -param {Object} [options]
   *
   * -throws {TypeError}
   *
   */

Question: should be used format:
   @param {string} title - The title of the book.
or 
  -param {String} target  
?

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+      this._id = `${this._params.call_id}${this._params.from_tag}${this._params.to_tag}`;
+
+      debug('add dialog id=', this._id);
+      this._ua.newDialog(this);
```
Do you means lib/Dialog.js ?

As I see it's INVITE-dialog
1. different dialogs states for INVITE dialog, for SUBSCRIBE dialog and for NOTIFY dialog.
2. receiveRequest for INVITE: check ACK etc...

Class Notifier and Subscriber implements minimal dialog interface, I mean
dialog.id                             to be added to JsSIP dialog table
dialog.receiveRequest()      to receive Request with the dialog id.

ikq commented on lib/Notifier.js:
```diff
@@ -0,0 +1,377 @@
+    }
+
+    this._params.cseq++;
+    this._ua.sendRequest(JsSIP_C.NOTIFY, this._target, this._params, headers, body,
```
I removed the code from UA and in lined it in Subscriber/Notifer

ikq commented:
Hi,

I mostly fixed all the issues (style, comments, variable names, etc)

Probably I should modify a state of Subscriber and Notifier from string to a number. Then it will be closer to your style.

I thought about dialog:
It would be nice if you added an abstract class dialog,
and invite-dialog, subscriber and notifier dialogs would be his descendants.

It could turns out something like:
```
module.exports = class AbstractDialog
{
     // members:
  constructor(){

     // to create outgoing & check incoming requests 
     this._state;
     this._remote_seqnum;
     this._local_seqnum;
     this._local_uri;
     this._remote_uri;
     this._remote_target;
     this._route_set;
 }

 // There should be methods to add/remove dialog 
// from stack dialogs table
  create(){
    this._ua.newDialog(this);
  }

  get id()
  {
    return this._id;
  }

  get state()
  {
    return this._state;
  }

  get local_seqnum()
  {
    return this._local_seqnum;
  }

  set local_seqnum(num)
  {
    this._local_seqnum = num;
  }

  terminate()
  {
  }

  sendRequest(method, options = {})
  {
  }

  receiveRequest(request)
  {
  }

  _createRequest(method, extraHeaders, body)
  {
  }

  _checkInDialogRequest(request)
  {
  }
};
```
IMHO Minimal dialog interface are:
- has id (string)
- can be added/removed to/from stack dialog table.
- receive incoming request  with the id.
- has state (number). 
   State set are different for each dialog, but always exists "terminated" state.

It's already implemented in Subscriber/Notifier.
(except that state is string)

other dialog functionality
- check incoming requests 
- keep set of parameters to create outgoing request.

Seems it's all according RFC 3261 and RFC 6665.

-----------------------------------------------------------

However, this is not at all necessary.

Rather for those who likes Object Oriented Programming.

I understand that design is always to the taste of the developer, so I don’t presume to advise something.

Everything works well in its present form.

jmillan commented:
Hi, 

Super busy lately. Will come back to this when we have some time.

jmillan commented:
H @ikq,

`Dialog` should be agnostic to the method. Why can't current `Dialog` implementation be used here? It has basically the `sendRequest()` method that you can call, and it will call the `receiveRequest()` method on your Subscriber/Notifier implementation.

ikq commented:
Hi José,

Dialog.js it's not abstract, but for INVITE dialog:

1.  constructor()
used INVITE dialog states: C.STATUS_EARLY : C.STATUS_CONFIRMED that missed in subscriber/notifier.

Here used different states.

2. this._ack_seqnum = this._remote_seqnum;
   not used in subscriber/notifier

3. receiveRequest() contain INVITE dialog code:
 if (request.method === JsSIP_C.ACK && this._ack_seqnum !== null)
    {
      this._ack_seqnum = null;
    }
    // INVITE received. Set this._ack_seqnum.
    else if (request.method === JsSIP_C.INVITE)

4. _createRequest()
(method === JsSIP_C.CANCEL || method === JsSIP_C.ACK) 

5. _checkInDialogRequest(request)
if (request.method === JsSIP_C.ACK)

etc...

Dialog.js is not usable for the case, because most of the 
methods and constructor should be overloading.

Dialog.js  code is very INVITE/ACK/CANCEL specific. 

IMHO If use OOP style, better create some base Dialog class.
and then extends it for subsclasses: for INVITE dialog, for subscriber and notifier dialogs.

ikq commented:
Hmm... I checked the Dialog code again.
sendRequest and receiveRequest probably can be used.

I just did not consider this possibility before because I found a lot of INVITE specific in the Dialog and I always try to do everything as simple as possible...

jmillan commented:
There is INVITE related code in Dialog because a dialog for INVITE method needs to behave in certain way, different to generic dialogs, but that's an implementation detail. Other clases should be able to create dialogs the same way RTCSession does.

ikq commented:
I will tests the subscriber/notifier version in the next week.

I cannot check authentication, because my server don't ask it.
there will be 2 tests: subscriber with notifier, and subscriber with some professional production server.

ikq commented:
Testing of the current version has been completed successfully

ikq commented:
subscribe authentication checked using special test.
IMHO: the subscriber/notifier  is ready to merge.

ikq commented:
I see I should change debug/debugerror to logger...

jmillan commented:
Hi @ikq,

We are super busy lately. Please keep using your branch and update this PR whenever you find any issue/enhancement.

Also, tests would accelerate the adoption.

ikq commented:
Тhis version has passed our testing (the version before the logger was changed) and we provide it to our customers.

Two browsers test can be taken from:
[https://github.com/ikq/subscribe_notify_test](https://github.com/ikq/subscribe_notify_test)

jmillan commented:
@ikq, more than an external test I mean tests in JsSIP as part of this PR, testing the clases created here.

N4COM commented:
Hi, is there some news? I'd like to test this functionality.

ikq commented:
> Hi, is there some news? I'd like to test this functionality.

IMHO it's stable code. 
[At least we use **subscriber** in our production, currently no issues reported]

The subscriber/notifier test you can taken from https://github.com/ikq/subscribe_notify_test

markusatm commented:
Hi, is there a specific reason why Subscriber and Notifier don't use the Contact header from the UA like RTCSession does? If the user provides a custom contact_uri in the UAConfiguration he would have to manually add it as a extra header to Subscriber and Notifier again, wouldn't he?

N4COM commented:
> > Hi, is there some news? I'd like to test this functionality.
> 
> IMHO it's stable code.
> [At least we use **subscriber** in our production, currently no issues reported]
> 
> The subscriber/notifier test can taken from https://github.com/ikq/subscribe_notify_test

I'd like to use the official repository, so I'd like if it will be merged ASAP

ikq commented:
> Hi, is there a specific reason why Subscriber and Notifier don't use the Contact header from the UA like RTCSession does? If the user provides a custom contact_uri in the UAConfiguration he would have to manually add it as a extra header to Subscriber and Notifier again, wouldn't he?

Yes, I'm not using the Configuration.contact_uri

The reason is purely historical. Subscriber was at first an autonomous module based on jssip client transaction.

Let check, how the configuration used for INVITE: 

UA.js:
   this._contact = {
      pub_gruu  : null,
      temp_gruu : null,
      uri       : this._configuration.contact_uri,

RTCSession.js:
  this._contact = this._ua.contact.toString({
      anonymous,
      outbound : true
    });


Thanks.
It can be improved (use by default Configuration.contact_uri)

mitre-mgarber commented:
@ikq if you don't mind external contributions i'd be willing to take a swing at writing unit tests for this project's test framework

ikq commented:
> f you don't mind external contributions i'd be willing to take a swing at writing unit tests for this project's test framework

**Thank you, I will gladly accept your help.**

I have provided an external test for 2 browsers: one is sending an SUBSCRIBE and the other is receiving it and sending a NOTIFY.

It's like checking a INVITE call with two browsers.
By the way, I did not find such INVITE unit test among the JsSIP tests:
So I don't know how it can be written with SUBSCRIBE/NOTIFY.

```
gulpfile.js:

gulp.task('test', function()
{
  // var src = 'test/*.js';
  const src = [
    'test/test-classes.js',
    'test/test-normalizeTarget.js',
    'test/test-parser.js',
    'test/test-properties.js',
    'test/test-UA-no-WebRTC.js',
    'test/test-digestAuthentication.js'
  ];

  return gulp.src(src)
    .pipe(expect(EXPECT_OPTIONS, src))
    .pipe(nodeunit({ reporter: 'default' }));
});
```

stefang42 commented:
> Thanks.
> It can be improved (use of Configuration.contact_uri)
> 
> P.S.
> Usage Configuration.contact_uri will modify default SUBSCRIBE contact - will be missed "+sip.instance" (that I add to default SUBSCRIBE Contact)
> User still can set "+sip.instance" to SUBSCRIBE Contact via custom Contact header.

@ikq Is there a specific reason to use Configuration.contact_uri? Why not just do it as in RTCSession, with `this._ua.contact.toString`? Then it would also use the gruu if one was provided. Or am I missing something?

ikq commented:
Thanks to markusatm to found the problem and to stefang42 proposed fix.
Now Contact in notifier and subscriber set as ua._contact.toString()

jmillan commented on lib/Notifier.js:
```diff
@@ -96,7 +96,7 @@ module.exports = class Notifier extends EventEmitter
     if (!this._contact)
     {
-      this._contact = `Contact: <sip:${subscribe.to.uri.user}@${Utils.createRandomToken(12)}.invalid;transport=ws>`;
+      this._contact = `Contact: ${this._ua._contact.toString()}`;
```
:+1:

jmillan commented:
Hi,

Yes, in order for this PR to be merged we need tests in JsSIP project.

You don't need two JsSIP instances, just one instance and you should inject a transport that you control (a FAKE transport) that you use to send JsSIP UA SUBSCRIBE|NOTIFY messages and receive and check what comes from JsSIP.

If you need some help please let us know.

ikq commented:
Hi, 
I added nodeunit test to test subscriber notifier communication.

1. Implemented loopSocket that send message itself, but modified Call-ID in each leg.
(without the modification method findDialog  confuses SUBSCRIBE dialog with NOTIFY dialog)

2. subscriber and notifier run in the same JsSIP instance and send/receive SIP messages.

The test SIP sequence is:
  - SUBSCRIBE
  - SUBSCRIBE OK
  - NOTIFY
  - NOTIFY OK
  - unSUBSCRIBE (with expires 0)
  - unSUBSCRIBE OK
  - final NOTIFY (with Subscription-State: terminated)
  - final NOTIFY OK.

The unit test work silently, but if uncomment line JsSIP.debug.enable('JsSIP:*');
can be seen complete SIP communication.
  
Some parts of SIP messages checked. 
Mostly bodies content and Content-Type. 
If you want can be added more assertions (checking) for some messages headers.

ikq commented:
> @ikq if you don't mind external contributions i'd be willing to take a swing at writing unit tests for this project's test framework

@mitre-mgarber
Thanks for the help offered.
Suddenly, I got the idea to study нodeunite testing, and it seemed that something began to work out.
However, let's see how it goes further.

ikq commented:
Hi,
Please check this unit test.
In my opinion it is ready, well, at least for a review.

jmillan commented:
Thanks a lot @ikq,

I'm quite busy at the moment but I'll take a look when I can.

jmillan commented:
Hi @ikq,

I don't see the reason for 'dialogCreated' event in Subscriber. Is there for historical/testing purposes?

ikq commented:
'dialogCreated' event was used to not send subsequence subscribes until will be response (with to tag) to initial SUBSCRIBE.

In the latest version of subscriber it automatically enqueues subsequences subscribes until the response to initial subscribe (event 'dialogCreated'), and then send them.

So you are right - now the event is not important.

ikq commented:
Another note about time calculation when next subscribe will be send.
Here used different from JsSIP registrar, time calculation.

Chrome since version 88 enter to intensive timer throttling mode: [https://developer.chrome.com/blog/timer-throttling-in-chrome-88/](url)

In the mode javascript timer works with resolution (time deviation) 1 minutes.
Even with such a bad timer, the subscriber will work.

[in my JsSIP clone I use the same for registrar].

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,592 @@
+        this._dialog = dialog;  
+
+        logger.debug('emit "dialogCreated"');
+        this.emit('dialogCreated');
```
Instead of this, we should fire an event like 'accepted'. Which represents that is happening.

ikq replied:
OK

ikq replied:
OK

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,592 @@
+        logger.debug('emit "dialogCreated"');
+        this.emit('dialogCreated');
+
+        // Subsequent subscribes saved in the queue until dialog created.
```
Instead of enqueuing. We should not send any subsequent subscribe until the initial was accepted. Hence the logic is placed in the application:

subscribe.on('accepted', () => { send another subscribe})

ikq replied:
I'm agree - it is good if the user writes code in this style.
So better have such event.

However, it is good also to have subsequence enqueue/dequeue code in subscriber.

For example, this will allow to call the subscribe.terminate() at any time (even before the event 'accepted')

E.g. user send initial subscribe, and after this suddenly want terminate subscription,
in the case he call subscriber.terminate() and don't think about events and state machine...

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,592 @@
+  terminate(body = null) 
+  {
+    logger.debug('terminate()');
+
```
Avoid calling this method multiple times. This should be the last public method to be called in the instance.

```suggestion
if (this._terminated)
{
  return;
}

this.terminated = true;
```

jmillan replied:
Better than the above would be checking:

`this._state === C.STATE_TERMINATED`

ikq replied:
It could be better, but ... not according RFC 6665.

subscriber switched to terminated state not after sending un-subscribe, but 
after receive terminated notify or after receive non ok response to subscribe.

See diagram: [https://datatracker.ietf.org/doc/html/rfc6665](url)
page 10.

So I used your "above" way:
To prevent duplicate terminate() call I used property **this._send_unsubscribe**
(instead this._terminated).
If you think that it's more clear I can change the name to **_terminated**.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,592 @@
+    this._send_unsubscribe = true;
+
+    // Set header Expires: 0.
+    const headers = this._headers.map((s) => 
+    { 
+      return s.startsWith('Expires') ? 'Expires: 0' : s; 
+    });
```
Cosmetic:

```suggestion
    const headers = this._headers.map((header) => 
    { 
      return header.startsWith('Expires') ? 'Expires: 0' : header; 
    });
```

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,592 @@
+
+  _scheduleSubscribe(expires) 
+  {
+    const timeout = expires >= 140 ? (expires * 1000 / 2) 
```
Please, add the corresponding log here indicating the rationale of this calculus.

ikq replied:
/*
      If expires less than 140 seconds we don't support Chrome high throttling timer mode. In the case re-SUBSCRIBE will send 5 seconds before expiration.
	  
	  Chrome high throttling mode: in the worst case timer fired 60 seconds later.
	  We give the server 10 seconds to make sure it will execute the command even if it is heavily loaded. So we order re-subscription time not later than 70 seconds before expiration. We also add randomization.
	  So time calculated as half time interval + (half interval - 70) * random.
	  
	  E.g. if expires is 140, re-subscribe will be ordered in 70 seconds.
	       if expires is 600, re-subscribe will be ordered in 300 + (0 .. 230) seconds.
	 */
  
I feel that explanation is too big to comment ;-)

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,371 @@
+    {
+      logger.warn(dialog.error);
+
+      throw new Error('SUBSCRIBE missed Contact');
```
Why are we throwing this specific error message?

Can't the Contact be checked before creating the dialog?

ikq replied:
Moved to begin of constructor()

```
    super();

    if (!subscribe)
    {
      throw new TypeError('subscribe is undefined');
    }

    if (!subscribe.hasHeader('contact'))
    {
      throw new TypeError('subscribe - no contact header');
    }

```

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,371 @@
+      return;
+    }
+
+    let expiresValue = request.getHeader('expires');
+
+    if (expiresValue === undefined || expiresValue === null) 
+    { 
+      // Missed header Expires. RFC 6665 3.1.1. Set default expires value.  
+      expiresValue = '900';
+      logger.debug(`Missed expires header. Set by default ${expiresValue}`);
+    }
+
+    this._expires = parseInt(expiresValue);
```
Cosmetic/Readability:

```suggestion    
    if (request.hasHeader('expires')) 
    { 
      this._expires = request.getHeader('expires')
    }
    else
    {    
      // RFC 6665 3.1.1, default expires value.
      this._expires = 900;
      
      logger.debug(`missing Expires header field, default value set: ${this._expires}`);
    }
```

ikq replied:
OK, 

One note - getHeader returns string, not number
So better use this._expires = parseInt(request.getHeader('expires'))

ikq commented:
Hi,
I made some changes.
But not finished yet (not checked '_terminated' issue)

ikq commented:
Hi,
I finished the changes after review.

jmillan commented:
Hi @ikq, I've been out for a while and couldn't check this. Let us please some more time to review it properly.

stefang42 commented:
Hi @ikq, I think it would be good to also add type information for the events to the type declaration files Subscriber.d.ts and Notifier.d.ts, those are currently missing. I mean like in RTCSession.d.ts, the RTCSessionEventMap and the various event handler declarations, as well as a declaration of the "on" method RTCSession.on(). Otherwise there is no type support when defining the event handlers in TypeScript.

stefang42 commented on lib/Subscriber.d.ts:
```diff
@@ -0,0 +1,22 @@
+declare enum SubscriberTerminatedCode {
+  SUBSCRIBE_RESPONSE_TIMEOUT = 0, 
+  SUBSCRIBE_TRANSPORT_ERROR = 1, 
+  SUBSCRIBE_NON_OK_RESPONSE = 2, 
```
This enum is not in sync with the corresponding termination code constants used in Subscriber.js, it is missing the value for SUBSCRIBE_BAD_OK_RESPONSE with value 3. Therefore all enum values here beginning with SUBSCRIBE_FAILED_AUTHENTICATION are off by 1, so event listeners for the 'terminated' event will interpret those codes incorrectly. E.g. when they receive the value 5 then according to this enum they will interpret it as "RECEIVE_FINAL_NOTIFY", but in fact value 5 is actually "UNSUBSCRIBE_TIMEOUT".

The enum should have these values:
```
declare enum SubscriberTerminatedCode {
  SUBSCRIBE_RESPONSE_TIMEOUT = 0, 
  SUBSCRIBE_TRANSPORT_ERROR = 1, 
  SUBSCRIBE_NON_OK_RESPONSE = 2,
  SUBSCRIBE_BAD_OK_RESPONSE = 3,
  SUBSCRIBE_FAILED_AUTHENTICATION = 4,
  UNSUBSCRIBE_TIMEOUT = 5,
  RECEIVE_FINAL_NOTIFY = 6,
  RECEIVE_BAD_NOTIFY = 7
}
```

ikq replied:
thanks stefang42. 
My problem that I don't use typescript. (so the part is unchecked).
If you can provide missed callbacks definition I can add it.
(other possibility wait until I learn the ts definitions ;-) or until the push request will be merged to master)

N4COM commented:
Hello, is there some news about this?

metinagaoglu commented:
Hello, is there any progress about this issue ?

ikq commented:
We have been using this code since august 2021. No customer complaints so far.
;-)
You welcome to use it if need.

metinagaoglu commented:
> We have been using this code since August 2021. (at the beginning the first versions and now the latest) No customer complaints so far. ;-) You welcome to use it if need.

We need this and probably use it. Thank you for your contribution.

SNH-48 commented:
find 2 promble
1.UA.d.ts export notifier that method name is wrong,it should be notify.
2.Constants.d.ts ALLOWED_METHODS miss SUBSCRIBE.

raiansantos commented:
Hey everyone! What is the pending items with this PR?

ikq commented:
Hi, I don't know why the code is still not merged. (Perhaps few people need it?)
We have been using this code since August 2021.

devoxy1 commented:
Hello! This is really good feature and we would like to use it. Is there chance this can be merged?

jmillan commented:
@ikq, I've re-reviewed the code and it's in good shape. The only missing part is the documentarion.  I'll prepare a new release tag so you can add them for the upcoming release.

jmillan commented:
> What's wrong with the documentation?
>  I thought you automatically generated it from comments.

Unfortunately not. We use a static website for the doc, and it's not automatically generated.

https://github.com/versatica/jssip.net

This has more than few years.., since JsSIP come up, and hence we are a bit outdated there.

I'll write a little README.md there with instructions on how to build and modify the docs.

jmillan commented:
A little README added to the JsSIP web repo. I've also created the 3.11.x version template branch  [here](https://github.com/versatica/jssip.net/tree/3.11.x).

The files to be modified are those under this folder:  content/documentation/3.11.x/

You will need to, at least, modify `api/ua.html` and create `api.subscriber.html` and `api.notifier.html` ones.

I know we should have autogenerated docs.., we would like to have the resources to do so.

Let us know if you find any issues.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,370 @@
+    this._terminated_reason = reason;
+    this._terminated_retry_after = retryAfter;
+
+    this.notify(body);
```
I don't like `this._terminated_reason` and `this._terminated_retry_after` being set here and making the public method `this.notify()` deal with them.

I would rather create a private method `_sendNotify(body, headers)`, that is used from `terminate()` and from the public `notify()`. Then `notify()` can assert that current state is not TERMINATED, otherwise throw.

ikq replied:
added private _notify() method that called from terminate() and notify().
notify() throw exception if called in TERMINATED state.

jmillan commented:
> Тhe tests are useful, because I myself have forgotten how to use this API.

I know, and I very much appreciate it but we need to have the public API documented in the website. I know by experience it's a hassle.., we'll try to move to autogenerated docs as we can.

jmillan commented:
@ikq, please, do not rely on the documentation branch I did for 3.11 version. I'm going to remove it right now.

We are going to have documented just the latest version, so after this `[PR](https://github.com/versatica/jssip.net/pull/14)` is merged, just clone it and add the docs there.

We definitely need to have the docs autogenerated from the inline comments. We already do it in [rtp.js](https://github.com/versatica/rtp.js/).

jmillan commented:
@ikq, please take merge `master` branch into this. I've added a new eslint rule `no-trailing-spaces`. There are many trailing spaces in this PR.

ikq commented:
In fact, I wrote the code a year and a half ago, and for some reason it seems was 
  interesting to me to allow the user *always* send a final notification with an 
  arbitrary body when the notifier dialog ends. 

  For the purpose it uses the sendFinalNotify parameter in the 'terminated' 
  event.
  this.emit('terminated', C.SUBSCRIPTION_EXPIRED, true);
 
  Currently this is only used for the case when the next SUBSCRIBE request does 
  not arrive before the expiration. C.SUBSCRIPTION_EXPIRED

This is a rare case, if the client did not send the subscribe on time, perhaps it does not work at all and it is not so important to send notifications with or without a body.

If for the case send automatically from timer callback final NOTIFY without body, 
sendFinalNotify argument can be removed from 'terminated' event and the user code will be a little simpler.

What do you think ?

ikq commented:
Previous commit was at Dec 7, 2021 - I forget how to use the API. 
Now I remember it using my own test
  https://github.com/ikq/subscribe_notify_test/blob/main/test.js

There was some idea how to a little simplify notifer API, but 
but after some thought, I decided that this doesn’t simplify much but reduces flexibility. So it's probably better to leave it as is.

devoxy1 commented:
@jmillan Would it be possible to review again? Thanks!

devoxy1 commented:
Up!

ddyner commented:
Hello. We also want to use this feature. Will this pull request be merged?

ikq commented:
I created pull request for https://github.com/versatica/jssip.net
Please take a look.

devoxy1 commented:
@jmillan Hello! Could you please check it up? Thank you!

jmillan commented:
I'll review when I can. If you are in a hurry, please do use this branch freely.

jmillan commented:
> I'll review when I can. If you are in a hurry, please do use this branch freely.

This way you will also collaborate on using, testing and reporting.

devoxy1 commented:
@jmillan We already using it this way :) but it's a bit difficult to maintain separate branch. 
Anyways, thanks you!

jmillan commented:
I'm not forgetting about this...,

devoxy1 commented:
@jmillan Hello! Any chance you had time to review this one?
Thanks!

orgads commented:
@jmillan Ping?

jmillan commented:
For those willing to use this feature I would suggest you use this branch. I don't have at the moment the resources to re-check and merge this, so please, do use this branch for the time being.

johnmarkovalo commented:
Looking forward with this merge also 👍

jmillan commented on lib/Config.js:
```diff
@@ -32,9 +32,10 @@ exports.settings = {
-  registrar_server : null,
+  register                : true,
+  register_expires        : 600,
+  register_from_tag_trail : '',
```
Everything related to `register_from_tag_trail` does not belong to this PR. It's already in master. Please remove it.

orgads replied:
Done

jmillan commented on lib/UA.d.ts:
```diff
@@ -40,6 +42,7 @@ export interface UAConfiguration {
   ha1?: string;
   register?: boolean;
   register_expires?: number;
+  register_from_tag_trail?: string | function() : string;
```
to be removed.

jmillan commented:
Also please do not edit `CHANGELOG.md`

jmillan commented on lib/RTCSession.js:
```diff
@@ -1988,16 +1993,15 @@ module.exports = class RTCSession extends EventEmitter
           connection.addEventListener('icegatheringstatechange', iceGatheringStateListener = () =>
           {
-            if ((connection.iceGatheringState === 'complete') && !finished)
+            if (connection.iceGatheringState === 'complete')
```
Change already in master, please remove.

orgads commented on lib/RTCSession.js:
```diff
@@ -1988,16 +1993,15 @@ module.exports = class RTCSession extends EventEmitter
             }
-
-            else if (!finished)
+            else
```
Done

jmillan commented:
Thanks, I'll take some time on a further, and hopefully last review.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    this._expires_timestamp = null;
+    this._expires_timer = null;
+
+    // Notifier state: pending, active, terminated. Not used: init, resp_wait.
```
The comment is unclear to me. What does this mean?:

>  Not used: init, resp_wait.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    this._state = pending ? C.STATE_PENDING : C.STATE_ACTIVE;
+
+    // Custom session empty object for high level use.
+    this.data = {};
```
For consistency, check how in `RtcSession.js` we use a `_data` member and then `get data()` and `set data()`.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    // Create dialog for normal and fetch-subscribe.
+    const dialog = new Dialog(this, subscribe, 'UAS');
+
+    this._dialog = dialog;
```
Remove unneeded temporal variable.

```suggestion
    this._dialog = new Dialog(this, subscribe, 'UAS');
```

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+
+    if (!subscribe.hasHeader('contact'))
+    {
+      throw new TypeError('subscribe - no contact header');
```
```suggestion
      throw new TypeError('Missing Contact header in subscribe request');
```

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+
+    if (!contentType)
+    {
+      throw new TypeError('contentType is undefined');
```
```suggestion
      throw new TypeError('Not enough arguments. Missing contentType');
```

jmillan replied:
Duplicated suggestion due to an error between GH and me.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    // Custom session empty object for high level use.
+    this.data = {};
+
+    this._dialog = null;
```
This line can be removed as dialog is created later in this method.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    else
+    {
+      // RFC 6665 3.1.1, default expires value.
+      this._expires = 900;
```
900 should better go into the corresponding constant value (`DEFAULT_EXPIRES_MS`) a the top of the file.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+  /**
+   * Private API
+   */
+  /**
```
```suggestion
   */
   
  /**
```

jmillan commented on lib/Notifier.d.ts:
```diff
@@ -0,0 +1,24 @@
+    NOTIFY_RESPONSE_TIMEOUT = 0,
+    NOTIFY_TRANSPORT_ERROR = 1,
+    NOTIFY_NON_OK_RESPONSE = 2,
+    NOTIFY_FAILED_AUTHENTICATION = 3,
```
```suggestion
    NOTIFY_AUTHENTICATION_FAILED = 3,
```

jmillan commented on lib/Notifier.d.ts:
```diff
@@ -0,0 +1,24 @@
+    NOTIFY_TRANSPORT_ERROR = 1,
+    NOTIFY_NON_OK_RESPONSE = 2,
+    NOTIFY_FAILED_AUTHENTICATION = 3,
+    SEND_FINAL_NOTIFY = 4,
```
```suggestion
    FINAL_NOTIFY_SENT = 4,
```

jmillan commented on lib/Notifier.d.ts:
```diff
@@ -0,0 +1,24 @@
+import {SUBSCRIBE, NOTIFY} from './Constants'
+import * as Utils from './Utils'
+
+declare enum NotifierTerminatedCode {
```
For consistency:


```suggestion
declare enum NotifierTerminationReason {
```

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+  RECEIVE_UNSUBSCRIBE          : 5,
+  SUBSCRIPTION_EXPIRED         : 6,
+
+  // Notifer states
```
```suggestion
  // Notifer states.
```

jmillan commented on lib/Subscriber.d.ts:
```diff
@@ -0,0 +1,23 @@
+  SUBSCRIBE_TRANSPORT_ERROR = 1,
+  SUBSCRIBE_NON_OK_RESPONSE = 2,
+  SUBSCRIBE_BAD_OK_RESPONSE = 3,
+  SUBSCRIBE_FAILED_AUTHENTICATION = 4,
```
```suggestion
  SUBSCRIBE_AUTHENTICATION_FAILED = 4,
```

jmillan commented on lib/Subscriber.d.ts:
```diff
@@ -0,0 +1,23 @@
+  SUBSCRIBE_FAILED_AUTHENTICATION = 4,
+  UNSUBSCRIBE_TIMEOUT = 5,
+  RECEIVE_FINAL_NOTIFY = 6,
+  RECEIVE_BAD_NOTIFY = 7
```
```suggestion
  WRONG_NOTIFY_RECEIVED = 7
```

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,605 @@
+  SUBSCRIBE_FAILED_AUTHENTICATION : 4,
+  UNSUBSCRIBE_TIMEOUT             : 5,
+  RECEIVE_FINAL_NOTIFY            : 6,
+  RECEIVE_BAD_NOTIFY              : 7,
```
same changes as suggested above.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,605 @@
+  {
+    if (response.status_code >= 200 && response.status_code < 300)
+    {
+      // Create dialog
```
Please add a dot at the end of all comments.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,605 @@
+
+        // RFC 6665 3.1.1 subscribe OK response must contain Expires header.
+        // Use workaround expires value.
+        expires_value = '900';
```
Should use the constant value (to be) defined at the top of the file.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,605 @@
+
+      if (expires_value !== 0 && !expires_value)
+      {
+        logger.warn('response without Expires header');
```
```suggestion
        logger.warn('response without Expires header, setting a default value of XYZ');
```

jmillan commented:
Currently UA fires `newSubscribe` with the incoming SUBSCRIBE request. 

If SUBSCRIBE request does not have a Contact header, the request WONT be responded, hence breaking the SIP protocol. Contact header is verified in Notifier constructor which simply throws if it's missing. This kind of sanity checks should be done in `Notifier.init_incoming(subscriber)` which should reject the request or emit `newSubscribe`. This method can be a static method in `Notifier`.

Check how it's done in RtcSession: https://github.com/versatica/JsSIP/blob/master/lib/RTCSession.js#L3499

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,387 @@
+    // Create dialog for normal and fetch-subscribe.
+    const dialog = new Dialog(this, subscribe, 'UAS');
+
+    this._dialog = dialog;
```
Duplicated suggestion due to an error between GH and me.

orgads commented:
I think all the comments should be now covered.

jmillan commented:
>  (I can't Resolve comments since I'm not the owner 🙄)

I will, no worries.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,409 @@
+   */
+  start()
+  {
+    logger.debug('start()');
```
Lets check  the current state here and throw if we are in an invalid one.

jmillan commented on lib/Notifier.js:
```diff
@@ -0,0 +1,409 @@
+      subsState += param;
+    }
+
+    let headers = this._headers.slice();
```
Please, always use `Utils.cloneArray(array)` rather than `array.slice()`

orgads replied:
Done. Pushed also #945 for another case.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+    if (parsed === -1)
+    {
+      throw new TypeError('eventName - wrong format');
+    }
```
This check should be moved to the beginning of the method with the other checks.

orgads replied:
Done

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+
+    request.reply(200);
+
+    const new_state = this._stateStringToNumber(subs_state.state);
```
This line can throw with the error "wrong state value", but one line above we already replied with 200. Let's move this line before `request.reply(200)` and negatively respond if the subscription state is not correct.

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+    logger.debug('terminate()');
+
+    // Prevent duplication un-subscribe sending.
+    if (this._terminated)
```
Why are we relying on `this._terminated` rather than checking `this._state = C.STATE_TERMINATED`? This member seems redundant.

orgads replied:
If I read it correctly, `_terminated` is meant to protect against multiple calls to `terminate`, and `_state` is only modified on `_terminateDialog`. Notice that the state is not modified here, and `_terminateDialog` is only called after a long timeout (30s).

jmillan commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+      // Check expires value.
+      let expires_value = response.getHeader('expires');
+
+      if (expires_value !== 0 && !expires_value)
```
Enhance readability.

```suggestion
      const expires = parseInt(expires_value);
      
      if (Utils.isDecimal(expires))
      {
         // Use it.
      }
      else
      {
        // Use the default value.
      }
```

jmillan commented:
Hi @orgads, I'm doing some modernisation on the library and I've moved code to `src/` rather than `lib/`. Please adapt the PR at your convenience.

I don't plan to do any other change affecting this PR.

orgads commented on lib/Notifier.js:
```diff
@@ -0,0 +1,409 @@
+      }
+
+      logger.debug(`emit "terminated" code=${C.SUBSCRIPTION_EXPIRED}, send final notify=true`);
+      this.emit('terminated', C.SUBSCRIPTION_EXPIRED, true);
```
It passes `true` for `sendFinalNotify`, and also doesn't call `this._dialog.terminate()`.

orgads commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+    // Check that arguments are defined.
+    if (!target)
+    {
+      throw new TypeError('target is undefined');
```
Done

orgads commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+      expires = C.DEFAULT_EXPIRES_MS;
+    }
+
+    this._expires = expires;
```
But I think it should also be set to the default value otherwise. Hopefully done as you meant.

orgads commented on lib/Subscriber.js:
```diff
@@ -0,0 +1,620 @@
+      this._headers.push(`Allow-Events: ${allowEvents}`);
+    }
+
+    // To enqueue subscribes created before receive initial subscribe OK.
```
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,418 @@
+    this._data = _data;
+  }
+
+  static init_incoming(request, callback)
```
Please move this before constructor. Static methods before member methods.

orgads replied:
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,418 @@
+   */
+  setActiveState()
+  {
+    logger.debug('setActiveState()');
```
```suggestion
    logger.debug('setActiveState()');
    
```

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,418 @@
+      }
+
+      logger.debug(`emit "terminated" code=${C.SUBSCRIPTION_EXPIRED}, send final notify=true`);
+      this.emit('terminated', C.SUBSCRIPTION_EXPIRED, true);
```
I wonder why the last argument is needed.

jmillan commented on src/Subscriber.d.ts:
```diff
@@ -0,0 +1,25 @@
+  get data(): any;
+  static get C(): typeof SubscriberTerminatedCode;
+  get C(): typeof SubscriberTerminatedCode;
+}
```
Same here, events are not typed.

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,418 @@
+    }, this._expires * 1000);
+  }
+
+  _stateToString()
```
For consistency? 

```suggestion
  _parseState()
```

jmillan commented on src/Subscriber.js:
```diff
@@ -0,0 +1,625 @@
+    if (this._state === C.STATE_INIT)
+    {
+      // fetch-subscribe - initial subscribe with Expires: 0.
+      this._sendInitialSubscribe(body, headers);
```
Is it even possible to call `terminate()` before `subscribe`? It should not. This method should throw if `this._state === C.STATE_INIT`.

orgads commented on src/Subscriber.js:
```diff
@@ -0,0 +1,625 @@
+
+    if (!this._headers.find((header) => header.startsWith('Contact')))
+    {
+
```
Done

orgads commented on src/Subscriber.js:
```diff
@@ -0,0 +1,628 @@
+
+    if (parsed === -1)
+    {
+      throw new TypeError('eventName - wrong format');
```
Done

orgads commented on src/Notifier.js:
```diff
@@ -0,0 +1,426 @@
+    return C;
+  }
+
+  get C()
```
Done

jmillan commented on src/Subscriber.js:
```diff
@@ -0,0 +1,629 @@
+    {
+      this._terminateDialog(C.UNSUBSCRIBE_TIMEOUT);
+    }, final_notify_timeout);
+  }
```
I don't think this logic is necessary.

When sending the request, the corresponding Non Invite Client Transaction will be created and held by `ua`. It will process the response or be terminated upon timeout.

The only thing to be done is to check if `this._state === C.STATE_TERMINATED` in `_receiveSubscribeResponse()`. If so, just return.

jmillan replied:
Meaning rather than having this timer logic, we can call just `this._terminateDialog()` right away.

jmillan replied:
Anyway please give a try. Check `RTCSession.js` upon sending `BYE` request.

orgads replied:
Done

jmillan commented:
>  Sorry, do you prefer to stack up commits instead?

I would prefer to see individual commits. This PR will be rebased before merging anyway.

orgads commented on src/Notifier.js:
```diff
@@ -0,0 +1,431 @@
+    });
+  }
+
+  _terminateDialog(termination_code)
```
Done

orgads commented on src/RTCSession.js:
```diff
@@ -18,6 +18,10 @@ const RTCSession_ReferNotifier = require('./RTCSession/ReferNotifier');
+/**
+ * @typedef {import('./SIPMessage.js').IncomingRequest} IncomingRequest
+ */
+
```
Type annotations. SipMessage is not imported. It is unrelated to this change though, just slipped it. I'll separate it.

jmillan commented on src/Subscriber.js:
```diff
@@ -393,13 +388,7 @@ module.exports = class Subscriber extends EventEmitter
-    {
-      this._terminateDialog(C.UNSUBSCRIBE_TIMEOUT);
-    }, final_notify_timeout);
+    this._terminateDialog(C.UNSUBSCRIBE_TIMEOUT);
```
I don't think the reason is correct. And probably such constant `UNSUBSCRIBE_TIMEOUT` should be removed.

jmillan replied:
I think now `this._terminated` makes no sense and that we can rely on `C.STATE_TERMINATED`, right?

jmillan replied:
I see that by terminating the dialog right away we are removing the possibility to receive a NOTIFY request with state `terminated` and a body..., mmm.

jmillan replied:
I think there is the need of a new state `WAITING_FOR_TERMINATED_NOTIFY`, which is set in this method, after calling `this._sendSubsequentSubscribe()`. If the SUBSCRIBE times out (no response is received) then we are already calling `this._terminateDialog()`. If we receive a NOTIFY with state `terminated` then we close it as we do now.

When calling `terminated()` we should check if the state is `STATE_TERMINATED` or `WAITING_FOR_TERMINATED_NOTIFY`.

orgads replied:
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,431 @@
+
+      return;
+    }
+    callback();
```
Since subscribe request is checked in multiple places, let's avoid redundant code.

```suggestion
  static init_incoming(request, callback)
  {
    try
    {
      Notifier.checkSubscribe(request);
    }
    catch (error)
    {  
      logger.debug('Notifier.init_incoming: invalid request');

      request.reply(405);

      return;
    }
    
    callback();
  }
    
 static checkSubscribe(subscribe)
 {
    if (!subscribe)
    {
      throw new TypeError('Not enough arguments. Missing subscribe request');
    }
    
    if (subscribe.method !== JsSIP_C.SUBSCRIBE)
    {
            throw new TypeError('Invalid method for Subscribe request');
    }

    if (!subscribe.hasHeader('contact'))
    {
      throw new TypeError('Missing Contact header in subscribe request');
    }

    if (!subscribe.hasHeader('event'))
    {
      throw new TypeError('Missing Event header in subscribe request');
    }  
  }
```

orgads replied:
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,431 @@
+    if (!eventName)
+    {
+      throw new TypeError('Missing Event header in subscribe request');
+    }
```
```suggestion
    Notifier.checkSubscribe(subscribe); 

    if (!contentType)
    {
      throw new TypeError('Not enough arguments. Missing contentType');
    }
```

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+      logger.warn('final notify already sent');
+
+      return;
+    }
```
Can we check state instead of dialog to be more explicit?

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+      if (!this._dialog)
+      {
+        return;
+      }
```
Also let's check state for explicitness.

orgads commented on src/Subscriber.js:
```diff
@@ -0,0 +1,622 @@
+    {
+      logger.warn(`Invalid Subscription-State header value: ${subs_state.state}`);
+      request.reply(400);
+      this._terminateDialog(C.WRONG_NOTIFY_RECEIVED);
```
Done

orgads commented on src/Subscriber.js:
```diff
@@ -0,0 +1,622 @@
+    {
+      return;
+    }
+    this._terminated = true;
```
Done

jmillan commented on src/Notifier.d.ts:
```diff
@@ -0,0 +1,31 @@
+  start(): void;
+  setActiveState(): void;
+  notify(body?: string): void;
+  terminate(body?: string, reason?: string): void;
```
```suggestion
  terminate(body?: string, reason?: string, retryAfter?: number): void;
```

orgads commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+
+    logger.debug(`emit "terminated" code=${termination_code}`);
+
+    this.emit('terminated', termination_code);
```
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+    {
+      // Set expires timer and time-stamp.
+      this._setExpiresTimer();
+    }
```
IMO this should be moved to `start()` after responding the SUBSCRIBE. Otherwise we may send the `terminated` NOTIFY before responding the SUBSCRIBE which is odd and probably not spec compliant.

jmillan reviewed (commented):
I think these are my last comments.

orgads commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+    subscribe.to_tag = Utils.newTag();
+
+    // Create dialog for normal and fetch-subscribe.
+    this._dialog = new Dialog(this, subscribe, 'UAS');
```
Done

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,443 @@
+    if (expires < 0)
+    {
+      expires = 0;
+    }
```
`private method` was misleading in my comment. Indeed the code suggestion is inlined.

jmillan commented on src/Notifier.js:
```diff
@@ -0,0 +1,469 @@
+      request.reply(405);
+
+      return;
+    }
```
Should be call `Notifier.checkSubscribe()` here instead?

jmillan replied:
Maybe not. We must be strict on the first SUBSCRIBE, but we can be lose on the others as we don't need such info.

jmillan commented:
> @ikq , @orgads I find the current status OK. Let me please ask you to go though this PR line by line once again and confirm that everything is fine.

Is this fine for you? I understand there's been a lot of work on both sides (mostly from your side). You implementing and me reviewing. I'm just asking for a last self-review and a ✅ from your part, indicating that you are OK and happy with the result, and commenting here anything that could be enhanced (even if it's in a future PR) or writing some test that could be good to have, if you feel so.

Let me know what you think about it.

jmillan commented:
Thank you @orgads @ikq for all the time taken on this!
